src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -152,13 +154,308 @@ using Parser = std::function<O(const Node&, DefaultRandom&)>;
 
 UniqueGenerator<int64_t> intGenerator(const Node& node, DefaultRandom& rng);
 UniqueGenerator<int64_t> int64GeneratorBasedOnDistribution(const Node& node, DefaultRandom& rng);
+UniqueGenerator<double> doubleGenerator(const Node& node, DefaultRandom& rng);
+UniqueGenerator<double> doubleGeneratorBasedOnDistribution(const Node& node, DefaultRandom& rng);
+UniqueGenerator<std::string> stringGenerator(const Node& node, DefaultRandom& rng);
+template <bool Verbatim, typename Out>
+Out valueGenerator(const Node& node,
+                   DefaultRandom& rng,
+                   const std::map<std::string, Parser<Out>>& parsers);
 
 template <bool Verbatim>
 std::unique_ptr<DocumentGenerator::Impl> documentGenerator(const Node& node, DefaultRandom& rng);
 
 template <bool Verbatim>
 UniqueGenerator<bsoncxx::array::value> arrayGenerator(const Node& node, DefaultRandom& rng);
 
+template <typename distribution,
```

Comment:
I realized I could template all this stuff rather than having a full class for every double generator. If I could figure out how to pass in a function to the template, I would generalize to NumberGeneraotr. I need to pass in doubleGenerator or intGenerator depending on the type. I suppose I could just put some special casing in for them.